arbnode/espresso_force_inclusion_checker.go

```diff
@@ -0,0 +1,155 @@
+type ForceInclusionChecker struct {
+	stopwaiter.StopWaiter
+
+	seqInbox              SeqInboxInterface
+	config                ForceInclusionCheckerConfig
+	l1Reader              *headerreader.HeaderReader
+	delayedMessageFetcher *DelayedMessageFetcher
+	fatalErrChan          chan error
+}
+
+func NewForceInclusionChecker(
+	seqInbox SeqInboxInterface,
+	config ForceInclusionCheckerConfig,
+	l1Reader *headerreader.HeaderReader,
+	delayedMessageFetcher *DelayedMessageFetcher,
+	fatalErrChan chan error,
+) *ForceInclusionChecker {
+	return &ForceInclusionChecker{
+		seqInbox:              seqInbox,
+		config:                config,
+		l1Reader:              l1Reader,
+		delayedMessageFetcher: delayedMessageFetcher,
+		fatalErrChan:          fatalErrChan,
+	}
+}
+
+func (f *ForceInclusionChecker) checkIfMessageCanBeForceIncluded(ctx context.Context) error {
+	// Get the total number of delayed messages read in the sequencer inbox
+	totalDelayedMessagesRead, err := f.seqInbox.TotalDelayedMessagesRead(ctx)
+	if err != nil {
+		return fmt.Errorf("error getting total delayed messages read: %w", err)
+	}
+
+	// Get the earliest block number that is without the force inclusion tolerance
+	badBlockNumber := f.getForceInclusionToleranceBlockNumber(ctx)
+	// Check the delayed message count at this block number
+	count, err := f.delayedMessageFetcher.getDelayedMessageCountAtBlock(badBlockNumber)
+	if err != nil {
+		return fmt.Errorf("error getting delayed message count at block %d: %w", badBlockNumber, err)
+	}
+	// If the message count in delay inbox is less than or equal to the total delayed messages read
+	// then no force inclusion is going to happen.
+	if count <= arbmath.BigToUintSaturating(totalDelayedMessagesRead) {
+		return nil
+	}
+	// Force inclusion is going to happen, panic the node.
+	err = fmt.Errorf("force inclusion is going to happen")
+	f.fatalErrChan <- err
+	return err
+}
+
+func (f *ForceInclusionChecker) Start(ctx context.Context) error {
+	f.StopWaiter.Start(ctx, f)
+
+	return f.CallIterativelySafe(func(ctx context.Context) time.Duration {
+		err := f.checkIfMessageCanBeForceIncluded(ctx)
+		if err != nil {
+			log.Error("error checking force inclusion", "err", err)
+			return f.config.RetryTime
+		}
+		return f.config.PollingInterval
+	})
+}
+
+func (f *ForceInclusionChecker) getForceInclusionToleranceBlockNumber(ctx context.Context) uint64 {
+	maxTimeVariationDelayBlocks, _, maxTimeVariationDelaySeconds, _, err := f.seqInbox.MaxTimeVariation(ctx)
+	if err != nil {
+		return 0
+	}
+	currentParentChainBlock, err := f.l1Reader.Client().BlockByNumber(ctx, nil)
+	if err != nil {
+		return 0
+	}
+
+	lastBadBlockNumber := arbmath.SaturatingUSub(f.config.BlockThresholdTolerance+currentParentChainBlock.NumberU64(), arbmath.BigToUintSaturating(maxTimeVariationDelayBlocks))
+	lastBadBlockTime := arbmath.SaturatingUSub(f.config.SecondThresholdTolerance+currentParentChainBlock.Time(), arbmath.BigToUintSaturating(maxTimeVariationDelaySeconds))
```

Comment:
Are we sure that these don't over/underflow? I see we are combining raw values from the config while doing some saturating subs. 
Just trying to make sure we cover our edge cases.